### PR TITLE
ARROW-417: Add Equals implementation to compare ChunkedArrays

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/column.h [hunk 1]
```diff
@@ -78,6 +80,9 @@ class ARROW_EXPORT Column {
 
   // @returns: the column's data as a chunked logical array
   std::shared_ptr<ChunkedArray> data() const { return data_; }
+
+  bool Equals(const std::shared_ptr<Column>& other) const;
```
Comment (wesm) [inline cpp/src/arrow/column.h]: Can you make the API for these instead `const ChunkedArray& other` and `const Column& other`? More flexible that way. we should try to do this consistently across the codebase. May make sense to maintain a wrapper function that takes `const shared_ptr<T>&` (sort of like `DataType::Equals`) that checks for nullptr like here

File: cpp/src/arrow/table.h [hunk 2]
```diff
@@ -100,6 +100,8 @@ class ARROW_EXPORT Table {
   // @returns: the number of rows (the corresponding length of each column)
   int64_t num_rows() const { return num_rows_; }
 
+  bool Equals(const std::shared_ptr<Table>& other) const;
```
Comment (wesm) [inline cpp/src/arrow/table.h]: Ditto with above comment.

File: cpp/src/arrow/column.h (hunk 1 above)
Comment (xhochy) [inline cpp/src/arrow/column.h]: Done.

File: cpp/src/arrow/table.h (hunk 2 above)
Comment (xhochy) [inline cpp/src/arrow/table.h]: Done

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**